### DISCUSSION TITLE
Options error: You must define TUN/TAP device (--dev)

### DISCUSSION THREAD

**Discussion by ryancasler** [haugene/docker-transmission-openvpn#2674] [Q&A] [answered]
Don't know what this means or how to fix it.  The config that I was using worked fine until I rebooted my host OS.  All of a sudden the container will no longer start.  Any assistance would be appreciated.  This is the command I am using to start my container:

```
docker run --cap-add=NET_ADMIN -d \
-v /mnt/m/downloads/:/data \
-v /mnt/m/vpn/config/:/config \
-e OPENVPN_PROVIDER=CUSTOM \
-v /mnt/m/vpn/scripts:/scripts \
-e TRANSMISSION_WEB_UI=shift \
-e OPENVPN_USERNAME=USERNAME \
-e OPENVPN_PASSWORD=PASSWORD \
-e LOCAL_NETWORK=192.168.1.0/24 \
-e TRANSMISSION_RATIO_LIMIT=0 \
-e TRANSMISSION_RATIO_LIMIT_ENABLED=true \
-e TRANSMISSION_SCRIPT_TORRENT_DONE_ENABLED=true  \
-e TRANSMISSION_SCRIPT_TORRENT_DONE_FILENAME=/scripts/complete.sh \
-e TRANSMISSION_SCRAPE_PAUSED_TORRENTS_ENABLED=false \
-e WEBPROXY_ENABLED=false  \
-e CREATE_TUN_DEVICE=true \
--log-driver json-file \
--log-opt max-size=10m \
--restart=unless-stopped \
--name transmission \
-p 9091:9091 \
-e TZ=America/New_York \
--pull=always \
-e PUID=1000 \
-e PGID=1000 \
haugene/transmission-openvpn
```

**Comment by pkishino** [comment]
The container isn’t and create the tun device.. check the logs more carefully.. what os?

**Reply by ryancasler** [reply]
Windows via Windows Subsystem for Linux.  Here is the log output:
```
Starting container with revision: 1103172c3288b7de681e2fb7f1378314f17f66cf
TRANSMISSION_HOME is currently set to: /config/transmission-home
WARNING: Deprecated. Found old default transmission-home folder at /data/transmission-home, setting this as TRANSMISSION_HOME. This might break in future versions.
We will fallback to this directory as long as the folder exists. Please consider moving it to /config/transmission-home
Creating TUN device /dev/net/tun
Using OpenVPN provider: CUSTOM
Modifying /etc/openvpn/custom/default.ovpn for best behaviour in this container
Modification: Point auth-user-pass option to the username/password file
Modification: Change ca certificate path
Modification: Change ping options
Modification: Update/set resolv-retry to 15 seconds
Modification: Change tls-crypt keyfile path
Modification: Set output verbosity to 3
Modification: Remap SIGUSR1 signal to SIGTERM, avoid OpenVPN restart loop
Modification: Updating status for config failure detection
Setting OpenVPN credentials...
adding route to local network 192.168.1.0/24 via 172.17.0.1 dev eth0
2023-07-31 20:18:11 Cipher negotiation is disabled since neither P2MP client nor server mode is enabled
Options error: You must define TUN/TAP device (--dev)
Use --help for more information.
```

**Comment by ryancasler** [comment]
I also removed this line from my default.opvn file "
`dev tun `
and it still didn't start

**Reply by edgd1er** [reply]
Two links that may help
Dev tun is required in config: https://forums.openvpn.net/viewtopic.php?t=25330

Wsl may be the cause: https://unix.stackexchange.com/questions/628542/wsl-vpn-connection-problem

**Reply by ryancasler** [reply]
Yeah, that is really no help at all because I don't know what any of those settings in the manual are or what they do or which ones to use.  Why would this have been working fine before and just stopped?

**Reply by ryancasler** [reply]
After looking at the files you linked to, there is nothing I can do to set up the tune device.  That has to be done in the command that launches openvpn.   So, unless I re-write the container, I can't fix this.

**Reply by pkishino** [reply]
Well, that is kinda the point of docker.. you can just destroy and recreate it anytime.. 
reasons why this might have worked before but not now:
- provider changed the server settings/profiles
- Windows updates 
- Docker updates 
- Possibly container updates (depends if you pulled the latest or nor, which image version are you using etc..) but I don’t recall any thing that would have changed the tun device stuff

**Reply by ryancasler** [reply]
So, basically, you're saying I'm on my own to figure it out?  Thanks for the help.

**Reply by pkishino** [reply]
Have you checked this?
https://haugene.github.io/docker-transmission-openvpn/tips-tricks/#running_it_on_a_nas

you should be able to manually create a tun device, mount this into the container and then set 
CREATE_TUN_DEVICE=false to test this..
If that works for you then we could look into why the create doesn’t work..
I don’t use windows nor am I technical support,
The expectation of using the software (or any foss really) is that we might be able to help if we have time etc 
You are expected to look through previous issues,  read the documentation, try and figure it out yourself as much as..

**Reply by ryancasler** [reply]
Yes, I tried that as well. I installed openvpn, made a connection, saw the tun device in /dev/net/tun, changed the parameters to build my container to use that device but I still got the same error when starting the container. I have tried everything you have suggested and nothing has worked.

**Reply by ryancasler** [reply]
I also tried on an Ubuntu VM and on a  RPi running Raspberry Pi OS and received the same error on every system I try it on.   Want to tell me again about how I have to try things more?   Did you look at the command I'm using to start the container to see if there was anything I was doing that was wrong?  I can also post the contents of my openvpn file if that could shed some light on this.   but when I use that file directly with open vpn I am able to connect no problem.

**Reply by ryancasler** [reply]
See, there is also this error:
CUSTOM provider specified but not using default.ovpn, will try to find a valid config mounted to /etc/openvpn/custom
which leads me to think that it is not using my default.ovpn file that I have mapped to in the first  volume mount.   Do I have to define that somewhere else too?

**Reply by pkishino** [reply]
I spun up two containers on macOS and Debian using your command.. are you sure that is your command? Where are you mounting the custom ovpn?

**Comment by ryancasler** [comment, marked as answer]
THAT'S ALL IT WAS!!!!!
adding 
"-v /mount/point/for/conig/:/etc/openvpn/custom \
to my docker run command fixed the problem.

**Reply by pkishino** [reply]
Good you fixed it.. I also missed this until I tried to run the command..

**Reply by edgd1er** [reply]
Good, now, you will have the opportunity to handle that warning, before an upcoming version removes that deprecated setting.
```
WARNING: Deprecated. Found old default transmission-home folder at /data/transmission-home, setting this as TRANSMISSION_HOME. This might break in future versions.
We will fallback to this directory as long as the folder exists. Please consider moving it to /config/transmission-home
```

**Reply by ryancasler** [reply]
Yes, I already fixed that one.  It is now located inside the config folder.

**Comment by SamJongenelen** [comment]
In my case, my .ovpn file got corrupted. Replacing it fixed this error for me
